File: reduce.f
      Program Example3d
       Implicit None
       Include 'mpif.h'
       Integer Size, Rank, Ierr, I, N, argc
       Integer*8 nPart
       Double Precision Sum, GSum, A, B, time1, time2, Al, Bl, X, F,
     $ timeT, TRUE, Yl, Yr, PI
       Character argv*32
       Parameter (PI = 3.14159265358979323846)                                                                                 
       Parameter (A=0.0, B=2.0*PI)
       Parameter (TRUE=-2.0*PI)  
       F(x)=(1-8*x*x)*cos(4*x)                                                                                                 

       Call MPI_INIT(Ierr)
       Call MPI_COMM_SIZE(MPI_COMM_WORLD, Size, Ierr)
       Call MPI_COMM_RANK(MPI_COMM_WORLD, Rank, Ierr)

       argc = iargc()
       CALL getarg(1,argv)
       READ(argv,*) nPart
       time1 = MPI_WTime()

       Al = A+(B-A)*Rank/Size
       Bl = Al+(B-A)/Size
       N = nPart/Size

       Sum = 0
       Yl = F(Al)
       Do I = 1,N
       X = Al+(Bl-Al)*I/N
       Yr = F(X)
       Sum = Sum + Yr + Yl
       Yl = Yr
       End Do

      
       Call MPI_REDUCE(Sum, GSum, 1, MPI_DOUBLE_PRECISION,
     $ MPI_SUM, 0, MPI_COMM_WORLD, Ierr)
       

       If (Rank.eq.0) Then
       time2 = MPI_WTime() 
       GSum = GSum/(N*Size)/2*(B-A)
       Write (6,*) 'Result= ',GSum,' Error= ',TRUE-GSum
       Write (6,*), ' Time = ', time2 - time1
       End If
       Call MPI_FINALIZE(Ierr)
       Stop
       End

File: reduce.cc
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <cmath>

#define TRUEVAL -2. * M_PI

const double A = 0;
const double B = 2*M_PI;

double func(double x) {return (1-8*x*x)*cos(4*x);}

int main(int argc, char** argv) {
  int size, rank;
  long long N = atoll(argv[1]);
  double Sum, GSum, time1, time2, timeI, Al, Bl, X, Yl, Yr, F, ISum, timeT;

  MPI_Init(NULL, NULL);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  time1 = MPI_Wtime();

  N /= size;
  Al = A+(B-A)*rank/size;
  Bl = Al+(B-A)/size;

  Sum = 0;
  Yl = func(Al);
  for (long long i = 0; i < N; ++i){
    X = Al+(Bl-Al)*(1+i)/N;
    Yr = func(X);
    Sum = Sum + Yr + Yl;
    Yl = Yr;
  }



  MPI_Reduce(&Sum, &GSum, 1, MPI_DOUBLE_PRECISION, MPI_SUM, 0, MPI_COMM_WORLD);
 
  if (rank == 0) {
    time2 = MPI_Wtime();
    GSum = GSum/(N*size)/2*(B-A);
    std::cout << "Result= " << GSum << " Error= " << (TRUEVAL-GSum) << std::endl <<
    " Time = " << time2 - time1 << std::endl;
  }

  MPI_Finalize();
  return 0;
}
